### PR TITLE
Update Scripting API to Store Context for Scripts

### PR DESCRIPTION
**Pull request by jdconrad**
Change stored scripts/file scripts to take in a context, and ensure that context is available at compile-time as a precursor to this feature (#20426).

This is currently a WIP since it's a large change, and there's some way to go.  I wanted to get early feedback on this portion.

The steps for this change are the following:
- [ ] Modify stored scripts API to allow for context, store lang, and make the namespace only based on id.  
- [ ] Modify the file script API to allow for context, store lang, and make the namespace only based on id.
- [ ] Update compilation by splitting the scripts into the 3 types of file, stored, and inline.  Ensure that context is always passed in upon compilation.

### CODE REVIEW

**Comment by nik9000** [comment]
I dropped `v5.0.0` because it isn't making that release. I dropped `breaking` and added `breaking-java` because I expect this to break plugins and maybe the transport client a little but we'll have to get it in in a backwards compatible way so that we can backport it to the 5.x branch.

File: core/src/main/java/org/elasticsearch/script/Script.java [hunk 1]
```diff
@@ -19,272 +19,936 @@
...
+                '}';
         }
-        return new Script(script, type, lang, params, contentType);
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(lang, params, script, type, contentType);
+    public interface ScriptBinding {
+        Object compile(ScriptEngineService engine, String id, String code, Map<String, String> options);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/script/Script.java]: Should this be `ScriptBinding<T>` where `T` is the interface that we expect the script to implement? Then `compile` can return `T`?

File: core/src/main/java/org/elasticsearch/script/Script.java [hunk 2]
```diff
@@ -19,272 +19,936 @@
...
+        public Object compile(ScriptEngineService engine, String id, String code, Map<String, String> options) {
+            return engine.compile(id, code, options);
+        }
     }
 
-    public interface ScriptField {
-        ParseField SCRIPT = new ParseField("script");
-        ParseField LANG = new ParseField("lang");
-        ParseField PARAMS = new ParseField("params");
+    public static final class SearchScriptBinding implements ScriptBinding {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/script/Script.java]: I feel like this should be somewhere in the search infrastructure rather.

File: core/src/main/java/org/elasticsearch/script/Script.java [hunk 3]
```diff
@@ -19,272 +19,936 @@
...
+        private SearchScriptBinding() {}
+
+        @Override
+        public Object compile(ScriptEngineService engine, String id, String code, Map<String, String> options) {
+            return engine.compile(id, code, options);
+        }
     }
 
+    public static final String DEFAULT_SCRIPT_LANG = "painless";
+    public static final String DEFAULT_SCRIPT_NAME = "<inline>";
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/script/Script.java]: Maybe rename to `INLINE_SCRIPT_NAME`?

File: core/src/main/java/org/elasticsearch/script/Script.java [hunk 4]
```diff
@@ -19,272 +19,936 @@
...
+
+    public static final String CONTENT_TYPE_OPTION = "content_type";
+
+    private Script() {}
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/script/Script.java]: Can you add a comment about why you need the empty private ctor?

**Review by nik9000 — COMMENTED** [review]
This has already gotten pretty large! I wonder if you can cut it into smaller PRs when it is closer to ready so we can review in smaller chunks.

**Comment by jdconrad** [comment]
I'm going to separate out this change into smaller pieces.  I'm going to close this for now.

**Pull request closed without merging**